openrtb/bid.go

```diff
@@ -1,31 +1,298 @@
+	//   communicated between buyer and exchanges a priori.
+	Tactic string `json:"tactic,omitempty"`
+
+	// Attribute:
+	//   cat
+	// Type:
+	//   string array
+	// Description:
+	//   IAB content categories of the creative. Refer to List 5.1.
+	Categories []Category `json:"cat,omitempty"`
+
+	// Attribute:
+	//   attr
+	// Type:
+	//   integer array
+	// Description:
+	//   Set of attributes describing the creative. Refer to List 5.3.
+	CreativeAttributes []CreativeAttribute `json:"attr,omitempty"`
+
+	// Attribute:
+	//   api
+	// Type:
+	//   integer
+	// Description:
+	//   API required by the markup if applicable. Refer to List 5.6.
+	API APIFramework `json:"api,omitempty"`
+
+	// Attribute:
+	//   protocol
+	// Type:
+	//   integer
+	// Description:
+	//   Video response protocol of the markup if applicable. Refer to
+	//   List 5.8.
+	// Pointer is not necessary for 0 is useless
+	Protocol Protocol `json:"protocol,omitempty"`
+
+	// Attribute:
+	//   qagmediarating
+	// Type:
+	//   integer
+	// Description:
+	//   Creative media rating per IQG guidelines. Refer to List 5.19.
+	// Pointer is not necessary for 0 is useless
+	QAGMediaRating IQGMediaRating `json:"qagmediarating,omitempty"`
+
+	// Attribute:
+	//   language
+	// Type:
+	//   string
+	// Description:
+	//   Language of the creative using ISO-639-1-alpha-2. The nonstandard
+	//   code “xx” may also be used if the creative has no
+	//   linguistic content (e.g., a banner with just a company logo).
+	Language string `json:"language,omitempty"`
+
+	// Attribute:
+	//   dealid
+	// Type:
+	//   string
+	// Description:
+	//   Reference to the deal.id from the bid request if this bid
+	//   pertains to a private marketplace direct deal.
+	DealID string `json:"dealid,omitempty"`
+
+	// Attribute:
+	//   w
+	// Type:
+	//   integer
+	// Description:
+	//   Width of the creative in device independent pixels (DIPS).
+	Width int `json:"w,omitempty"`
+
+	// Attribute:
+	//   h
+	// Type:
+	//   integer
+	// Description:
+	//   Height of the creative in device independent pixels (DIPS).
+	Height int `json:"h,omitempty"`
+
+	// Attribute:
+	//   wratio
+	// Type:
+	//   integer
+	// Description:
+	//   Relative width of the creative when expressing size as a ratio.
+	//   Required for Flex Ads.
+	WRatio int `json:"wratio,omitempty"`
+
+	// Attribute:
+	//   hratio
+	// Type:
+	//   integer
+	// Description:
+	//   Relative height of the creative when expressing size as a ratio.
+	//   Required for Flex Ads.
+	HRatio int `json:"hratio,omitempty"`
+
+	// Attribute:
+	//   exp
+	// Type:
+	//   integer
+	// Description:
+	//   Advisory as to the number of seconds the bidder is willing to
+	//   wait between the auction and the actual impression.
+	Exp int `json:"exp,omitempty"`
+
+	// Attribute:
+	//   ext
+	// Type:
+	//   object
+	// Description:
+	//   Placeholder for bidder-specific extensions to OpenRTB
+	Extension json.RawMessage `json:"ext,omitempty"`
```

Comment:
I noticed some object defines the extension is an interface, some object defines it as RawMessage. Are we able to make them unified?

Reply:
Yes, we can. But let's talk more before make decision.
Bid uses RawMessage but Video uses interface, because the former belongs to BidResponse and Jaeger only need to parse its extension, and the latter belongs to BidRequest and Jaeger only need to make its extension.
It's convenient to use interface as extension to make it and use RawMessage as extension to parse it. On the contrary, that will be troublesome. So, current design is convenient for SSP, but a nightmare for DSP.
If we want to make vungo openrtb really valuable for most applications, there are several possible solutions and no clear winner.